### DISCUSSION TITLE
Middleware global environment is different from server environment

### DISCUSSION THREAD

**Discussion by IvanKuzyshyn** [vercel/next.js#47951] [Help]
### Preconditions:
1. I have a Next.js using a custom server based on Express.
2. I have a service which implements interface like this:
```typescript
interface ExampleService {
  init(): Instance
  getInstance(): Instance
  shutdown(): Promise<void>
}
```
Let's assume that service implementation is similar to this example:
```typescript
const id = Symbol.for('example-service') // unique id

init() {
  global[id] = { ... } // instance of the service, in real life it is like: new ExampleService(...)
  return global[id] 
}
getInstance() {
  return global[id] 
}
 shutdown() {
  return global[id].shutdown() 
 }
```
This service is used to validate the request header value.
3. Service instance is created on Express server bootstrap and available globally 
```typescript
async function bootstrap() {
  init() // initialize service
  const nextServer = next({dev: isDev, port, hostname})
  await nextServer.prepare()
  const expressServer = express()
  // other code...
  expressServer.listen(config.port, ...);
}
```
4. And this means that I can get access to the instance like this:
```typescript
// get instance in getServerSideProps 
export const getServerSideProps: GetServerSideProps = async (context) => {
  const exampleService = getInstance(); // I have access to global instance
}
// get instance in API handler
export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  const exampleService = getInstance(); // I have access to global instance
}
```
5. But for some reason, I could not get access to this instance in middleware.
```typescript
// middleware.ts
export default async function middleware(request: NextRequest) {
 const exampleService = getInstance(); // returns undefined
}
```
For me, it seems that the middleware executing environment is different from the server execution environment. Did I get it right? Is it because it is running on edge runtime?

Question:
- if the statement above is correct how to share instances between server execution context and middleware? Is there any best practice or workaround? I could not initialize a new instance of the same service in the middleware because it is required to do `instance.shutdown()` when the service is not used anymore.
- if the statement above is not correct what is the potential issue in my implementation?

**Comment by DogeAmazed** [comment]
We see the same behaviour when using `global` (e.g. for a `NodeCache` instance). It seems like `middleware` and `getServerSideProps` aren't sharing these, and as you said, running in isolated environments. Both `middleware` and `getServerSideProps` are running on the Node runtime, so it's not just a Edge behaviour.

It seems like this isn't even documented?

**Comment by leerob** [comment]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental
